MSAL/src-internal/ios/MSALLegacySharedAccountsProvider.m

```diff
@@ -0,0 +1,512 @@
+        }
+        
+        if ([account matchesParameters:parameters])
+        {
+            [resultAccounts addObject:account];
+        }
+    }
+    
+    return resultAccounts;
+}
+
+#pragma mark - Update
+
+- (BOOL)updateAccount:(id<MSALAccount>)account idTokenClaims:(NSDictionary *)idTokenClaims error:(NSError **)error
+{
+    MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, nil, @"Updating account %@", MSID_PII_LOG_MASKABLE(account));
+    
+    return [self updateAccount:account
+                 idTokenClaims:idTokenClaims
+                tenantProfiles:nil
+                     operation:MSALLegacySharedAccountUpdateOperation
+                         error:error];
+}
+
+- (nullable NSArray<MSALLegacySharedAccount *> *)updatableAccountsFromJsonObject:(NSDictionary *)jsonDictionary
+                                                                     msalAccount:(id<MSALAccount>)msalAccount
+                                                                   idTokenClaims:(NSDictionary *)idTokenClaims
+                                                                         version:(MSALLegacySharedAccountVersion)version
+                                                                           error:(NSError **)error
+{
+    MSALAccountEnumerationParameters *parameters = [MSALLegacySharedAccountFactory parametersForAccount:msalAccount tenantProfileIdentifier:idTokenClaims[@"oid"]];
+    
+    if (!parameters)
+    {
+        NSError *parameterError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unsupported account found, skipping update", nil, nil, nil, nil, nil);
+        [self fillAndLogError:error withError:parameterError logLine:@"Unsupported account found, skipping update"];
+        return nil;
+    }
+    
+    NSError *parseError = nil;
+    NSArray<MSALLegacySharedAccount *> *accounts = [self accountsFromJsonObject:jsonDictionary withParameters:parameters error:&parseError];
+    
+    if (parseError)
+    {
+        [self fillAndLogError:error withError:parseError logLine:@"Failed to parse accounts"];
+        return nil;
+    }
+    
+    if (![accounts count])
+    {
+        NSError *accountError = nil;
+        MSALLegacySharedAccount *sharedAccount = [MSALLegacySharedAccountFactory accountWithMSALAccount:msalAccount
+                                                                                                 claims:idTokenClaims
+                                                                                        applicationName:self.applicationIdentifier
+                                                                                         accountVersion:version
+                                                                                                  error:&accountError];
+        if (!sharedAccount)
+        {
+            if (accountError)
+            {
+                [self fillAndLogError:error withError:accountError logLine:@"Failed to create account"];
+                return nil;
```

Comment:
nit: Shouldn't we either return an empty array or nil.

Reply:
We generally follow this practice:
empty array == I didn't find anything, but no error encountered
nil == I had an error during lookup